manta-pay/src/crypto/hash/poseidon/parameter_generation/round_numbers.rs

```diff
@@ -0,0 +1,140 @@
+// Copyright 2019-2022 Manta Network.
+// This file is part of manta-rs.
+//
+// manta-rs is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// manta-rs is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with manta-rs.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Generate number of full & partial rounds
+//! Adapted from <https://github.com/filecoin-project/neptune/blob/master/src/round_numbers.rs>
+
+// The number of bits of the Poseidon prime field modulus. Denoted `n` in the Poseidon paper
+// (where `n = ceil(log2(p))`). Note that BLS12-381's scalar field modulus is 255 bits, however we
+// use 256 bits for simplicity when operating on bytes as the single bit difference does not affect
+// the round number security properties.
+const PRIME_BITLEN: usize = 256;
+
+// Security level (in bits), denoted `M` in the Poseidon paper.
+const M: usize = 128;
+
+/// The number of S-boxes (also called the "cost") given by equation (14) in the Poseidon paper:
+/// `cost = t * R_F + R_P`.
+fn n_sboxes(t: usize, rf: usize, rp: usize) -> usize {
+    t * rf + rp
+}
+
+/// Returns the round numbers for a given arity `(R_F, R_P)`.
+pub fn round_numbers_base(arity: usize) -> (usize, usize) {
+    let t = arity + 1;
+    calc_round_numbers(t, true)
+}
+
+/// In case of newly-discovered attacks, we may need stronger security.
+/// This option exists so we can preemptively create circuits in order to switch
+/// to them quickly if needed.
+///
+/// "A realistic alternative is to increase the number of partial rounds by 25%.
+/// Then it is unlikely that a new attack breaks through this number,
+/// but even if this happens then the complexity is almost surely above 2^64, and you will be safe."
+/// - D Khovratovich
+pub fn round_numbers_strengthened(arity: usize) -> (usize, usize) {
+    let (full_round, partial_rounds) = round_numbers_base(arity);
+
+    // Increase by 25%, rounding up.
+    let strengthened_partial_rounds = f64::ceil(partial_rounds as f64 * 1.25) as usize;
+
+    (full_round, strengthened_partial_rounds)
+}
+
+/// Returns the round numbers for a given width `t`. Here, the `security_margin` parameter does not
+/// indicate that we are calculating `R_F` and `R_P` for the "strengthened" round numbers, done in
+/// the function `round_numbers_strengthened()`.
+pub fn calc_round_numbers(t: usize, security_margin: bool) -> (usize, usize) {

```

Comment:
Dead code? Round numbers are hardcoded in `config.rs`, and this file is no longer used.

Reply:
This still tells us what the safety margin is. We should add a test to check that the values defined in the spec match this (since we can't run this function as a `const-expr` yet).

Reply:
This still tells us what the safety margin is. We should add a test to check that the values defined in the spec match this (since we can't run this function as a `const-expr` yet).

Reply:
I added a test for that.